docs/RFCS/20211207_graceful_draining.md

```diff
@@ -0,0 +1,422 @@
+   is confusing. We propose to rename this function to
+   `s.sqlServer.isHealthcheckReady`.)
+
+2. New SQL connections and queries are not allowed from the start of
+   `query_wait`.
+
+3. The draining process is reported periodically in `cockroach-data/logs/cockroach.log`,
+   but it does not shows the draining phases (e.g. `query_wait`).
+
+4. If all SQL connections are idle or there are no SQL connections during
+   `drain_wait`, the server still does a hard wait till the end of `drain_wait`.
+
+5. If all SQL connections become query-free at a timestamp within `query_wait`,
+   the server performs an early exit.
+
+6. If a user acquires a connection before `query_wait` starts, but sends the
+   server a query on this connection after `query_wait` starts, this user will
+   receive `ERROR: server is shutting down`.
+
+7. Throughout the whole draining process, HTTP connections are still allowed.
+   The restrictions are only for *SQL* connection.
+
+In summary, the features of the two periods in the current draining process,
+`drain_wait` and `query_wait`, are as below:
+![Summary of features for drain_wait and query_wait][summary pic]
+To better illustrate the behavior of the current draining process, and the
+problem it brings along, we made an experiment with the [HirakiCP][HirakiCP]
+connection pool is built [here][demo code],
+and a demo recording can be found [here][demo recording current draining].
+
+# Technical Design
+([A draft PR can be found here.][draft pr])
+
+This design proposes to optimize the draining process by adding a new
+waiting period between `drain_wait` and `query_wait` called `connection_wait`,
+during which the server waits for clients to close SQL connections.
+New SQL connections are not allowed, and any SQL connections that are not pulled
+by the user and expire the `MaxLifeTime` will be closed by the connection pool.
+I.e. **no more client queries will be allowed to this draining node within this
+period, and the connection pools are expected to recycle idle SQL connections.**
+
+If at a certain timestamp all SQL connections are closed, the server performs an
+early exit and proceeds to `lease_transfer_wait`. Note that if a connection is
+pulled by a user and waiting for query input, this connection is not considered
+idle, and will not be recycled until the user puts it back in the pool. The
+timeout is finite. Clients can notice from the log that
+the existing SQL connections that are not executing queries will be closed.
+The following is the proposed draining process with optimized waiting stages,
+and early exit.
+
+![Proposed draining process][proposed draining pic]
+
+Most of the changes should be made in the server and pgwire package.
+Similar to existent `drain_wait` and `query_wait`, the new wait period will be
+declared as a `*settings.DurationSetting` as follows:
+
+```go
+connectionWait = settings.RegisterDurationSetting(
```

Comment:
cluster settings now need to marked as allowed or disallowed for multitenant. i think this one should be "TenantReadOnly" (but not tenant writable)